Sources/CvPlayerAI.cpp

```diff
@@ -12852,9 +12852,58 @@ int CvPlayerAI::AI_unitPropertyValue(UnitTypes eUnit, PropertyTypes eProperty) c
 	return iValue;
 }
 
-/********************************************************************************/
-/* 	City Defenders						24.07.2010				Fuyu			*/
-/********************************************************************************/
+int CvPlayerAI::AI_unitBuildingValue(UnitTypes eUnit, CvArea* pArea) const
+{
+	FAssertMsg(eUnit != NO_UNIT, "Unit is not assigned a valid value");
+	FAssertMsg(eUnitAI != NO_UNITAI, "UnitAI is not assigned a valid value");
+
+	CvUnitInfo& kUnitInfo = GC.getUnitInfo(eUnit);
+
+	// Evaluate unit's ability to construct buildings
+	int iCount = 0;
+
+	if (kUnitInfo.getNumBuildings() > 0 && pArea != NULL && !isNPC())
+	{
+		int iBuildingValue;
+		int iLoop;
+		bool bCoastal = kUnitInfo.getDomainType() == DOMAIN_SEA;
+		int iMinOceanSize;
+		if (bCoastal)
+		{
+			iMinOceanSize = GC.getMIN_WATER_SIZE_FOR_OCEAN();
+		}
+
+		for (int iI = 0; iI < kUnitInfo.getNumBuildings(); iI++)
+		{
+			BuildingTypes eBuilding = (BuildingTypes)kUnitInfo.getBuildings(iI);
+
+			if (NO_BUILDING != eBuilding)
+			{
+				if (canConstruct(eBuilding, false, false, true) && AI_getNumBuildingsNeeded(eBuilding, bCoastal) > 0)
+				{
+					iCount = 0;
+					iBuildingValue = 0;
+					for (CvCity* pLoopCity = firstCity(&iLoop); pLoopCity != NULL; pLoopCity = nextCity(&iLoop))
+					{
+						if ((!bCoastal || pLoopCity->isCoastal(iMinOceanSize)) && (pLoopCity->area() == pArea))
+						{
+							if (pLoopCity->getNumBuilding(eBuilding) == 0 && pLoopCity->canConstruct(eBuilding, false, false, true))
```

Comment:
These can be collapsed into a single if statement instead of two nested ones. The || and && operators only evaluate their right hand side if they need to. 

e.g.
`if (true || expensiveFunction()) {}` will NEVER call `expensiveFunction()` because it can't change the result of the expression. Regardless of what `expensiveFunction()` returns the if statement would always be executed, because `true || true` is `true` and `true || false` is also `true`.

`if(false && expensiveFunction()) {} `will NEVER call `expensiveFunction()` because it also can't change the result. Regardless of what it returns the if statement will never be executed because `false && true` is `false` and `false && false` is also `false`.

The benefits of reducing it to a single statement are readability (deep nesting reduces readability) and expressiveness: 
It accurately expresses that you require all these conditions to be met to increase iCount, and that you are *not* implying that there could be other possibilities when the outer if statement is true but the inner one is false. i.e. It tells future programmers that they should not be considering adding an else statement to the inner if statement to handle where it fails.

Reply:
hmm... I think the thinking on this is that the first set of checks relate to each other and you could put a stop at each layer of related checks to more easily determine where a check set is failing if you feel it should be succeeding or succeeding where you feel it should be failing or simply helping to determine a smaller segment of possible reasons for failure of reaching the final result to a more limited selection of checks by not trying to stuff all the checks into one IF statement line.  So I'm not sure that I agree that deep nesting reduces readability - in fact I feel its very much the opposite in that it makes it even harder to read and harder to narrow in on a culprit spot during debug analysis.
It might come down to what you're used to seeing I suppose.  Is there a distinction much in the processing efficiency?

Mind you, I didn't invent this method of multiple nesting of if statements but I've found it very helpful when debugging and figured I should mention how.

Reply:
> So I'm not sure that I agree that deep nesting reduces readability

It's pretty much 100% agreed on in the programming community at large. [Here](https://en.wikibooks.org/wiki/Computer_Programming/Coding_Style/Minimize_nesting) is one of the many explanations you can find. 
This specific case it isn't so bad, as it all fits on a few lines. However I have seen plenty of places in the code where if statements are nested 3+ deep, and this is well understood to be bad for readability and maintainability.

> Is there a distinction much in the processing efficiency?

There is possibly a distinction in how well the optimizer can analyse it. Generally the more semantically concise your code is the better the optimizer can reason about it. This goes for people reading the code as well. If you see a single if statement you know that it models a single condition that the inner block relies on. If you see nested if statements you now need to consider the context of each one. Does it have an else statement, does it have code that is executed for one of the outer blocks but not the inner one? It is a less precise way of expressing your intent, and requires the reader and optimizer to look at all the code before they can conclude that you actually just intend this to be a single condition with a single piece of code executed if it passes.

> Mind you, I didn't invent this method of multiple nesting of if statements but I've found it very helpful when debugging and figured I should mention how.

Nobody invented it, it's just a typical naive coding style. i.e. It's something you learn is bad from having read and maintained a lot of code bases in various styles. Most coders will, at a glance, tell you that this code base is poorly written, and the deep nesting of statements will probably be the thing they are basing that on.

The "autos" window will help you debug if statements, it shows the return values of all functions and useful variables so you can see how an if statement was evaluated. 
Another option with complicated if statements is to consider encapsulating the condition in a function. This can make the code a lot more readable because the function name simply describes the condition you are testing, and the reader doesn't need to know what exactly the function does.